File: app/schemas/job_application.py
"""Pydantic schemas for JobApplication (lightweight job CRM)."""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

JOB_STATUSES = {"applied", "interview", "offer", "rejected"}


class JobApplicationCreate(BaseModel):
    company: str = Field(min_length=1, max_length=255)
    role: str = Field(min_length=1, max_length=255)
    status: str = Field(default="applied", pattern="^(applied|interview|offer|rejected)$")
    job_description: str | None = Field(None, max_length=20000)
    notes: str | None = Field(None, max_length=10000)
    applied_date: date | None = None
    follow_up_date: date | None = None
    job_posting_url: str | None = Field(None, max_length=2000)
    resume_version_id: uuid.UUID | None = None
    cover_letter_id: uuid.UUID | None = None


class JobApplicationUpdate(BaseModel):
    company: str | None = Field(None, min_length=1, max_length=255)
    role: str | None = Field(None, min_length=1, max_length=255)
    status: str | None = Field(None, pattern="^(applied|interview|offer|rejected)$")
    job_description: str | None = Field(None, max_length=20000)
    notes: str | None = Field(None, max_length=10000)
    applied_date: date | None = None
    follow_up_date: date | None = None
    job_posting_url: str | None = Field(None, max_length=2000)
    resume_version_id: uuid.UUID | None = None
    cover_letter_id: uuid.UUID | None = None


class JobApplicationRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    company: str
    role: str
    status: str
    job_description: str | None
    notes: str | None
    applied_date: date | None
    follow_up_date: date | None
    job_posting_url: str | None
    resume_version_id: uuid.UUID | None
    cover_letter_id: uuid.UUID | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class JobApplicationListResponse(BaseModel):
    items: list[JobApplicationRead]
    total: int
    offset: int
    limit: int

File: frontend/src/lib/types/job-application.ts
/** Mirrors FastAPI JobApplication schemas. */

export type JobStatus = "applied" | "interview" | "offer" | "rejected";

export const JOB_STATUS_LABELS: Record<JobStatus, string> = {
  applied: "Applied",
  interview: "Interview",
  offer: "Offer",
  rejected: "Rejected",
};

export const JOB_STATUS_COLORS: Record<JobStatus, string> = {
  applied: "bg-blue-500/15 text-blue-400",
  interview: "bg-yellow-500/15 text-yellow-400",
  offer: "bg-green-500/15 text-green-400",
  rejected: "bg-red-500/15 text-red-400",
};

export type JobApplication = {
  id: string;
  user_id: string;
  company: string;
  role: string;
  status: JobStatus;
  job_description: string | null;
  notes: string | null;
  applied_date: string | null;
  resume_version_id: string | null;
  cover_letter_id: string | null;
  created_at: string;
  updated_at: string;
};

export type JobApplicationCreate = {
  company: string;
  role: string;
  status?: JobStatus;
  job_description?: string | null;
  notes?: string | null;
  applied_date?: string | null;
  resume_version_id?: string | null;
  cover_letter_id?: string | null;
};

export type JobApplicationUpdate = Partial<JobApplicationCreate>;

export type JobApplicationListResponse = {
  items: JobApplication[];
  total: number;
  offset: number;
  limit: number;
};
